src/Ssh/Ssh/Common/SshBaseCmdlet.cs

```diff
@@ -412,44 +422,115 @@ protected internal void SetResourceType()
             }
 
             var resourcetypefilter = supportedResourceTypes.Select(type => $"resourceType eq '{type}'").ToArray();
-            String filter = $"$filter=name eq '{Name}' and ({String.Join(" or ", resourcetypefilter)})";
+            string filter = $"$filter=name eq '{Name}' and ({String.Join(" or ", resourcetypefilter)})";
             ODataQuery<GenericResourceFilter> query = new ODataQuery<GenericResourceFilter>(filter);
 
-            String[] types;
             try
             {
                 IPage<GenericResource> resources = ResourceManagementClient.Resources.ListByResourceGroupWithHttpMessagesAsync(ResourceGroupName, query).GetAwaiter().GetResult().Body;
-                types = resources.Select(resource => resource.Type).ToArray();
+
+                if (resources == null || !resources.Any())
+                {
+                    throw new AzPSResourceNotFoundCloudException(String.Format(Resources.ResourceNotFoundNoTypeProvided, Name, ResourceGroupName));
+                }
+                if (resources.Count() > 1)
+                {
+                    throw new AzPSArgumentException(String.Format(Resources.MultipleResourcesWithSameName, Name, ResourceGroupName), ResourceType);
+                }
+
+                GenericResource resource = resources.First();
+
+                SetResourceType(resource);
+                SetAzureTargetResourceTags(resource);
+
             }
             catch (CloudException exception)
             {
                 throw new AzPSCloudException(String.Format(Resources.ListResourcesCloudException, ResourceGroupName, exception.Message));
             }
-            catch (ArgumentNullException)
+        }
+
+        protected internal void SetResourceType(GenericResource resource)
+        {
+            if (ParameterSetName.Equals(IpAddressParameterSet))
             {
-                throw new AzPSApplicationException(String.Format(Resources.ListResourcesArgumentNullException, ResourceGroupName));
+                ResourceType = "Microsoft.Compute/virtualMachines";
+                return;
             }
 
-            if (ResourceType != null)
+            if (ResourceType != null && !ResourceType.Equals(resource.Type, StringComparison.CurrentCultureIgnoreCase))
             {
-                if (!types.Contains(ResourceType, StringComparer.CurrentCultureIgnoreCase))
+                throw new AzPSResourceNotFoundCloudException(String.Format(Resources.ResourceNotFoundTypeProvided, Name, ResourceType, ResourceGroupName));
+            }
+
+            ResourceType = resource.Type;
+        }
+
+        protected internal void SetAzureTargetResourceTags(GenericResource resource)
```

Comment:
Since this method is calling ConfigurePortNumberFromResourceTag, this is doing more than just setting the resource tags. The name should reflect what the method is doing. Also, since only the Resource Tags property is only being used to set the port, maybe there is no need to have a Class Property for it, instead you could just pass that as a parameter to the ConfigurePortNumberFromResourceTag. I think you could have a protected internal method called ConfigurePortNumberFromResourceTag that calls a private method GetAzureTargetResourceTags that returns a dictionary of tags.